src/prpy/tsr/tsr.py

```diff
@@ -231,14 +336,113 @@ def from_yaml(x, *args, **kw_args):
         x_dict = yaml.safe_load(x, *args, **kw_args)
         return TSR.from_dict(x_dict)
 
-    def sample(self):
-        if len(self.TSRs) == 0:
-            return None
+    def is_valid(self, xyzrpy, ignoreNAN=False):
+        """
+        Checks if a xyzrpy list is a valid sample from the TSR.
+        @param xyzrpy a list of xyzrpy values
+        @param ignoreNAN (optional, defaults to False) ignore NaN xyzrpy
+        @return True if valid and False if not
+        """
+
+        if len(xyzrpy) != len(self.TSRs):
+            raise('Sample must be of equal length to TSR chain!')
+
+        for idx in range(len(self.TSRs)):
+            if not self.TSRs[idx].is_valid(xyzrpy[idx], ignoreNAN):
+                return False
+
+        return True
+
+    def to_transform(self, xyzrpy):
+        """
+        Converts a xyzrpy list into an
+        end-effector transform.
+
+        @param  a list of xyzrpy values
+        @return trans 4x4 transform
+        """
 
+        if not self.is_valid(xyzrpy):
+            raise ValueError('Invalid xyzrpy')
         T0_w = self.TSRs[0].T0_w
         for idx in range(len(self.TSRs)):
             tsr_current = self.TSRs[idx]
             tsr_current.T0_w = T0_w
-            T0_w = tsr_current.sample()
+            T0_w = tsr_current.to_transform(xyzrpy[idx])
 
         return T0_w
+
+    def sample_xyzrpy(self, xyzrpy=None):
+        """
+        Samples from Bw to generate an xyzrpy sample
+        Can specify some values optionally as NaN.
```

Comment:
This comment is inaccurate. This returns a list of `xyzrpy` samples, where the `i`-th entry corresponds to the values for the `i`-th TSR in the chain.